### ISSUE TITLE
Access Violation after running for several minutes 

### ISSUE THREAD

**Issue by rssk** [jangxx/node-win-desktop-duplication#1]
Thanks for the great library!  Unfortunately when running on win 10 with either `getFrame` `getFrameasync` or `startAutoCapture` Node's crashing after a couple minutes with: `Stop reason: Exception 0xc0000005 encountered at address 0x7ffb91a210e4: Access violation writing location 0x00000000`

Which seems to be happening here: https://github.com/jangxx/node-win-desktop-duplication/blob/master/src/desktopduplication.cpp#L320 which you think wouldn't happen given the successful `HRESULT`....?

**Comment by jangxx** [comment]
Apparently I should check if `malloc` returns a null pointer, which it seems to do in your case. I'm guessing you're running out of memory after several minutes, which causes malloc to fail. This could be caused by a memory leak introduced by me, or maybe you're simply not letting go of the image buffers, which could also cause the memory to fill up.

In any case, I need to add that return value check. Are you by any chance accumulating image buffers in your code? If you are, try letting the buffers go out of scope if you don't need them anymore, so the GC can free the buffers again. If you are not, I'll have to take another look at the library and hunt for the memory leak somewhere in there.

**Comment by rssk** [comment]
Good call, that is exactly what's happening. I've stripped down my code to this bare example that's failing for me.
```
let dd = new DesktopDuplication(0);

try {
  dd.initialize();
} catch(err) {
  console.log('An error occurred:', err.message);
  process.exit(0);
}

setInterval(() => {
  const frame = dd.getFrame();
}, 50);
```

**Comment by jangxx** [comment]
In that case, I need to take another look. The GC should free those buffers, but apparently it doesn't.

**Comment by rssk** [comment]
I'm definitely not an expert on node addons, but from what I can understand it seems like the addon and V8 dont share heap/stack and the return values are just copied over to V8's memory space, still leaving the original object around perhaps? 

It seems like the canonical solution for data transfer between spaces is using a shared Buffer [like here](https://stackoverflow.com/questions/28951978/pass-node-js-buffer-to-c-addon) and [here](https://community.risingstack.com/using-buffers-node-js-c-plus-plus/) as apparently that is much faster, which I think would be possible as long as you could know the screen size and format beforehand (h * w * rgba)?

**Comment by jangxx** [comment]
Huh, maybe I should learn to read or something. I was already using shared memory between my addon and Node, and I even supplied a `finalizer`, which was supposed to be called by the GC in case memory runs low. That was not happening however, at least not automatically. This problem was already described in the documentation for the node-addon-api however:

>The Napi::Buffer object does not assume ownership for the data and expects it to be valid for the lifetime of the object. Since the Napi::Buffer is subject to garbage collection this overload is only suitable for data which is static and never needs to be freed.

Simply copying the data to Node's heap and freeing the original buffer seems to fix the problem for me and I have pushed a version 1.0.1 with the fix. I would be happy if you could try it out and tell me, if it also fixes the problem on your end. If it does, I'll put it on npm.

**Comment by rssk** [comment]
Awesome, I really appreciate the quick response. 
This runs much longer, however now node longer exits but hangs when using `getFrameAsync` in an interval.

Typically this error output: ```Error: Failed to aquire next frame: The application made a call that is invalid. Either the parameters of the call or the state of some object was incorrect.
Enable the D3D debug layer in order to see details via debug messages.``` is the last thing before the interval hangs, but that happens often anyway, so I'm not sure if it's relevant or not.

I'm going to do some more testing and try to get some more to go on, but let me know if you have any ideas:-)

**Comment by rssk** [comment]
I think I've got a bit more to go on:

Its hanging with error with this error on `wrap_frame`'s `free` call:
```Stop reason: Exception 0xc0000005 encountered at address 0x7ffe464d134c: Access violation reading location 0xffffffffffffffff```
And here's stack:
```static bool check_bytes(const unsigned char *, const unsigned char, const unsigned __int64) (@7ffe464d134c..7ffe464d137b:3)
static bool is_block_an_aligned_allocation(const void *) (@7ffe464d27dd..7ffe464d282b:3)
static void free_dbg_nolock(void *, const int) (@7ffe464d1860..7ffe464d18f3:3)
_free_dbg (@_free_dbg:25)
free (@free:11)
public: class Napi::Value __cdecl DesktopDuplication::wrap_getFrame(class Napi::CallbackInfo const & __ptr64) __ptr64 (c:\users\user\projects\node-win-desktop-duplication\src\desktopduplication.cpp:382)
public: class Napi::Value __cdecl <lambda_7ba37853aceb9bdfdc1546e01af1f0c0>::operator()(void)const __ptr64 (c:\users\user\projects\node-win-desktop-duplication\node_modules\node-addon-api\napi-inl.h:3369)
struct napi_value__ * Napi::details::WrapCallback<class <lambda_7ba37853aceb9bdfdc1546e01af1f0c0> >(class <lambda_7ba37853aceb9bdfdc1546e01af1f0c0>) (c:\users\user\projects\node-win-desktop-duplication\node_modules\node-addon-api\napi-inl.h:69)
struct napi_value__ * Napi::ObjectWrap<class DesktopDuplication>::InstanceMethodCallbackWrapper(struct napi_env__ *,struct napi_callback_info__ *) (c:\users\user\projects\node-win-desktop-duplication\node_modules\node-addon-api\napi-inl.h:3362)\\
```

The frame itself has `RESULT_ERROR` but unfortunately my debugger is having issues telling me what the error string is exactly

**Comment by jangxx** [comment]
I'm pretty sure both of these problems are dealt with in my last commit. Try pulling the repo again.

The error you are seeing in `getFrameAsync` is due to the way the DesktopDuplication API is designed. I added an explanation in the troubleshooting section of the readme. If you want to get frames in an interval, using the auto capture thread is probably your best bet, since it's designed for exactly that purpose.

The second problem is probably due to a missing return statement, but I fixed that in the same commit.
